ipaserver/plugins/baseuser.py

```diff
@@ -694,3 +703,158 @@ class baseuser_remove_principal(LDAPRemoveAttribute):
+        issuer='<I>{}'.format(issuer) if issuer else '',
+        subject='<S>{}'.format(subject) if subject else '',
+        )
+
+
+def _convert_options_to_certmap(options):
+    """
+    Converts options to ipacertmapdata
+
+    When --subject --issuer or --usercertificate options are used,
+    the value for ipacertmapdata is built from extracting subject and issuer,
+    converting their values to X500 ordering and using the format
+    X509:<I>issuer<S>subject
+    For instance:
+    X509:<I>O=DOMAIN,CN=Certificate Authority<S>O=DOMAIN,CN=user
+    A list of values can be returned if --usercertificate is used multiple
+    times, or in conjunction with --subject --issuer.
+    """
+    data = []
+    for item in options.get('ipacertmapdata', ()):
+        data.append(item)
+
+    if 'issuer' in options or 'subject' in options:
+        issuer = options.get('issuer')
+        subject = options.get('subject')
+        data.append(_build_mapdata(subject, issuer))
+
+    for dercert in options.get('usercertificate', ()):
+        cert = x509.load_certificate(dercert, x509.DER)
+        issuer = DN(cert.issuer)
+        subject = DN(cert.subject)
+        data.append(_build_mapdata(subject, issuer))
+
+    return data
+
+
+class baseuser_add_certmap(LDAPAddAttributeViaOption):
+    __doc__ = _("Add one or more certificate mappings to the user entry.")
+    msg_summary = _('Added certificate mappings to user "%(value)s"')
+
+    attribute = 'ipacertmapdata'
+    takes_options = certmap_options
+
+    def get_options(self):
+        # ipacertmapdata is not mandatory as it can be built
+        # from the values subject+issuer or from reading usercertificate
+        for option in super(baseuser_add_certmap, self).get_options():
+            if option.name in ['ipacertmapdata']:
+                yield option.clone(required=False, alwaysask=False)
+            else:
+                yield option.clone()
+
+    def pre_callback(self, ldap, dn, entry_attrs, attrs_list, *keys,
+                     **options):
+        # The 3 valid calls are
+        # --subject xx --issuer yy
+        # --certificate
+        # --data
+        # Check that at least one of the 3 formats is used
+        valid_options = (
+            'issuer', 'subject', 'ipacertmapdata', 'usercertificate')
+        if all(key not in options for key in valid_options):
+            raise errors.RequirementError(name='data')
+
+        # The objectclass ipacertmapobject may not be present on
+        # existing user entries. We need to add it if we define a new
+        # value for ipacertmapdata
+        if 'objectclass' not in entry_attrs:
+            entry_attrs_old = ldap.get_entry(dn, ['objectclass'])
+        objclasses_lc = [x.lower() for x in entry_attrs_old['objectclass']]
+        if 'ipacertmapobject' not in objclasses_lc:
+            entry_attrs['objectclass'] = ['ipacertmapobject']
+
+        entry_attrs[self.attribute] = _convert_options_to_certmap(options)
```

Comment:
IMO it would be better to implement this using `default_from` on the `ipacertmapdata`, `issuer` and `subject` parameters.